packages/dds/tree/src/test/feature-libraries/default-schema/schemaChecker.spec.ts

```diff
@@ -0,0 +1,698 @@
+	};
+}
+
+function getValueNode(nodeSchemaIdentifier: string, value: Value): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value,
+		fields: new Map(),
+	};
+}
+function getMapNode(nodeSchemaIdentifier: string, fields: Map<FieldKey, MapTree[]>): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value: undefined,
+		fields,
+	};
+}
+
+function getObjectNode(nodeSchemaIdentifier: string, fields: Map<FieldKey, MapTree[]>): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value: undefined,
+		fields,
+	};
+}
+
+describe.only("schema validation", () => {
+	const multiplicityTestCases: [
+		kind: Multiplicity,
+		numberToTest: number,
+		expectedResult: boolean,
+	][] = [
+		[Multiplicity.Forbidden, 0, true],
+		[Multiplicity.Forbidden, 1, false],
+		[Multiplicity.Single, 0, false],
+		[Multiplicity.Single, 1, true],
+		[Multiplicity.Single, 2, false],
+		[Multiplicity.Sequence, 0, true],
+		[Multiplicity.Sequence, 1, true],
+		[Multiplicity.Sequence, 2, true],
+		[Multiplicity.Optional, 0, true],
+		[Multiplicity.Optional, 1, true],
+		[Multiplicity.Optional, 2, false],
+	];
+	for (const [kind, numberToTest, expectedResult] of multiplicityTestCases) {
+		it(`compliesWithMultiplicity(${numberToTest}, ${Multiplicity[kind]}) === ${expectedResult}`, () => {
+			const actual = compliesWithMultiplicity(numberToTest, kind);
+			assert.equal(actual, expectedResult);
+		});
+	}
+
+	describe("isNodeInSchema", () => {
+		describe("LeafNodeStoredSchema", () => {
+			const numberNode = getValueNode("myNumberNode", 1);
+			const stringNode = getValueNode("myStringNode", "string");
+			const booleanNode = getValueNode("myBooleanNode", false);
+			const nullNode = getValueNode("myNullNode", null);
+			const undefinedNode = getValueNode("myUndefinedNode", undefined);
+			const fluidHandleNode = getValueNode("myFluidHandleNode", new TestFluidHandle());
+			const nodeCases = [
+				numberNode,
+				stringNode,
+				booleanNode,
+				nullNode,
+				fluidHandleNode,
+				undefinedNode,
+			];
+
+			// Making the key of the record a ValueSchema ensures that we'll get compile-time errors if we add new
+			// ValueSchema values but forget to add test cases for them.
+			const testCases: Record<ValueSchema, any> = {
+				[ValueSchema.Number]: {
+					schema: new LeafNodeStoredSchema(ValueSchema.Number),
+					positiveNodeType: numberNode,
+				},
+				[ValueSchema.String]: {
+					schema: new LeafNodeStoredSchema(ValueSchema.String),
+					positiveNodeType: stringNode,
+				},
+				[ValueSchema.Boolean]: {
+					schema: new LeafNodeStoredSchema(ValueSchema.Boolean),
+					positiveNodeType: booleanNode,
+				},
+				[ValueSchema.Null]: {
+					schema: new LeafNodeStoredSchema(ValueSchema.Null),
+					positiveNodeType: nullNode,
+				},
+				[ValueSchema.FluidHandle]: {
+					schema: new LeafNodeStoredSchema(ValueSchema.FluidHandle),
+					positiveNodeType: fluidHandleNode,
+				},
+			};
```

Comment:
Thats a neat type safety trick, I like it, but I have an even safer way to do it (I think).

The way you have this loses the strong typing of the value (uses any). If you want this type checking, but without losing the strong types, you can use typescript's new satisfies keyword.

```suggestion
			// Making the key of the record a ValueSchema ensures that we'll get compile-time errors if we add new
			// ValueSchema values but forget to add test cases for them.
			const testCases = {
				[ValueSchema.Number]: {
					schema: new LeafNodeStoredSchema(ValueSchema.Number),
					positiveNodeType: numberNode,
				},
				[ValueSchema.String]: {
					schema: new LeafNodeStoredSchema(ValueSchema.String),
					positiveNodeType: stringNode,
				},
				[ValueSchema.Boolean]: {
					schema: new LeafNodeStoredSchema(ValueSchema.Boolean),
					positiveNodeType: booleanNode,
				},
				[ValueSchema.Null]: {
					schema: new LeafNodeStoredSchema(ValueSchema.Null),
					positiveNodeType: nullNode,
				},
				[ValueSchema.FluidHandle]: {
					schema: new LeafNodeStoredSchema(ValueSchema.FluidHandle),
					positiveNodeType: fluidHandleNode,
				},
			} satisfies Record<ValueSchema, unknown>;
```

Reply:
Thanks! Applied.